helper/common/common.go

```diff
@@ -99,22 +102,137 @@ func DirectoryExists(directoryPath string) bool {
 	return true
 }
 
-// createDir creates a file system directory if it doesn't exist
-func createDir(path string) error {
-	_, err := os.Stat(path)
+// Checks if the file at the specified path exists
+func FileExists(filePath string) bool {
+	// Grab the absolute filepath
+	pathAbs, err := filepath.Abs(filePath)
+	if err != nil {
+		return false
+	}
+
+	// Check if the file exists, and that it's actually a file if there is a hit
+	if fileInfo, statErr := os.Stat(pathAbs); os.IsNotExist(statErr) || (fileInfo != nil && fileInfo.IsDir()) {
+		return false
+	}
+
+	return true
+}
+
+// Creates a directory at path and with perms level permissions.
+// If directory already exists, owner and permissions are verified.
+func CreateDirSafe(path string, perms fs.FileMode) error {
+	info, err := os.Stat(path)
+	// check if an error occurred other than path not exists
 	if err != nil && !os.IsNotExist(err) {
 		return err
 	}
 
-	if os.IsNotExist(err) {
-		if err := os.MkdirAll(path, os.ModePerm); err != nil {
+	// create directory if it does not exist
+	if !DirectoryExists(path) {
+		if err := os.MkdirAll(path, perms); err != nil {
 			return err
 		}
+
+		return nil
+	}
+
+	// verify that existing directory's owner and permissions are safe
+	err = verifyFileAndTryUpdatePermissions(path, info, perms)
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
+// Creates a file at path and with perms level permissions.
+// If file already exists, owner and permissions are verified.
+// If shouldNotExist is true, an error is returned if file exists
+func CreateFileSafe(path string, data []byte, perms fs.FileMode, shouldNotExist bool) error {
+	info, err := os.Stat(path)
+	// check if an error occurred other than path not exists
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	fileExists := FileExists(path)
+	if fileExists && shouldNotExist {
+		return fmt.Errorf("%s already initialized", path)
+	}
+
+	// create file if it does not exist
+	if !fileExists {
+		if err := os.WriteFile(path, data, perms); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// verify that existing file's owner and permissions are safe
+	err = verifyFileAndTryUpdatePermissions(path, info, perms)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Verifies the file using `verifyFileOwnerAndPermissions` function
+// and updates permissions if current user is the owner
+func verifyFileAndTryUpdatePermissions(path string, info fs.FileInfo, expectedPerms fs.FileMode) error {
+	err, isCurrUserOwner := verifyFileOwnerAndPermissions(path, info, expectedPerms)
+	if err != nil {
+		return err
+	}
+
+	// update permissions (just in case) if current user is the owner of the directory
+	if isCurrUserOwner {
+		err := os.Chmod(path, expectedPerms)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Is there any reason why you fell you need this just in case block? Sound more risky to update if everything is fine rather than leaving just as it is.

Reply:
If the user wants to upgrade polygon-edge version, already created files and folders will exist with higher permissions. Or maybe the current user just wanted to play around or whatever and created folders manually with higher or lower permissions. In such cases the app should fix it if possible.

Reply:
done